clients/pkg/promtail/targets/file/filetarget.go

```diff
@@ -319,24 +319,56 @@ func (t *FileTarget) startTailing(ps []string) {
 			continue
 		}
 
-		level.Debug(t.logger).Log("msg", "tailing new file", "filename", p)
-		tailer, err := newTailer(t.metrics, t.logger, t.handler, t.positions, p, t.encoding)
-		if err != nil {
-			level.Error(t.logger).Log("msg", "failed to start tailer", "error", err, "filename", p)
-			continue
+		var reader Reader
+		if isCompressed(p) {
+			level.Debug(t.logger).Log("msg", "reading from compressed file", "filename", p)
+			decompresser, err := newDecompresser(t.metrics, t.logger, t.handler, t.positions, p, t.encoding)
+			if err != nil {
+				level.Error(t.logger).Log("msg", "failed to start decompresser", "error", err, "filename", p)
+				continue
+			}
+			reader = decompresser
+		} else {
+			level.Debug(t.logger).Log("msg", "tailing new file", "filename", p)
+			tailer, err := newTailer(t.metrics, t.logger, t.handler, t.positions, p, t.encoding)
+			if err != nil {
+				level.Error(t.logger).Log("msg", "failed to start tailer", "error", err, "filename", p)
+				continue
+			}
+			reader = tailer
 		}
-		t.tails[p] = tailer
+		t.readers[p] = reader
 	}
 }
 
+func isCompressed(p string) bool {
+	ext := filepath.Ext(p)
+
+	supportedCompressedFormats := []string{
+		".zip",
+		".gz",
+		".tar.gz",
+		".z",
+		".bz2",
+	}
+
+	for _, format := range supportedCompressedFormats {
+		if ext == format {
+			return true
+		}
+	}
```

Comment:
Micro-optimization: make `supportedCompressedFormats` a map to save a few unnecessary cycles here.

Reply:
Done by https://github.com/grafana/loki/pull/6708/commits/4bedfb27850286122f667e0fbb1a480aee627223. However, I added it as a function instead, since Go doesn't support constant maps. I'm also reusing it to list the supported extensions.